---
author: raoenhui
layout: post
title: "慎用try catch"
date: 2018-12-16 22:00
categories : Js
comments: true
tags:
- Js
---

### 前言
自从ECMA-262第3版引入了`try catch`语句，作为JavaScript中处理异常的一种标准方式。基本的语法如下所示。

### 一、try catch基本语法


```javascript
try {
    //可能会导致错误的代码
} catch (error) {
    //在错误发生时怎么处理
}finally {
     //即使报错始终执行
 }
```

### 二、try catch特点

##### 1.try catch耗性能

##### 1.1 try catch耗性能原理
[ECMAScript 2015 -The try Statement](https://www.ecma-international.org/ecma-262/6.0/#sec-try-statement-static-semantics-varscopeddeclarations)

[13.15.5](https://www.ecma-international.org/ecma-262/6.0/#sec-try-statement-static-semantics-vardeclarednames "link to this section") Static Semantics: VarDeclaredNames
- TryStatement : try Block Catch Finally
- 1.Let names be VarDeclaredNames of Block.
- 2.Append to names the elements of the VarDeclaredNames of Catch.
- 3.Append to names the elements of the VarDeclaredNames of Finally.
- 4.Return names.

[13.15.6](https://www.ecma-international.org/ecma-262/6.0/#sec-try-statement-static-semantics-varscopeddeclarations "link to this section") Static Semantics: VarScopedDeclarations
- TryStatement : try Block Catch Finally
- 1.Let declarations be VarScopedDeclarations of Block.
- 2.Append to declarations the elements of the VarScopedDeclarations of Catch.
- 3.Append to declarations the elements of the VarScopedDeclarations of Finally.
- 4.Return declarations.

根据上面[ECMAScript](https://www.ecma-international.org/ecma-262/6.0)`文档的`13.15.5`和`13.15.6`。

**下面仅为本妹子自己的翻译理解，仅供参考**

 上面大概说运行`try catch`时，需要将当前的词法环境和作用域全部分别添加到`catch`和`Finally`所要执行的代码块中。从上可以推断出`try catch`是消耗性能的。

##### 1.2 try catch耗性能实验
![webwxgetmsgimg.jpeg](http://upload-images.jianshu.io/upload_images/9902136-1df0bee220c7aadb?imageMogr2/auto-orient/strip)

下面我用`Chrome62`和`IE9`分别添加多个`try catch`,进行对比实验，虽然，很想抛弃`万恶的IE`，但是很多国内的产品不答应呀，除非我们去健身房再多练练，打一架，嘿嘿～～
1.2.1 实验数据：
```javascript
//没有加try catch
(function () {
  var i = 0;
     i++;
}())
```
```javascript
//有try catch
(function () {
  var i = 0;
    try {
      i++;
    } catch (ex) {
    } finally {
    }
}())
```
1.2.2 实验结果：
![chrome62-simple.png](http://upload-images.jianshu.io/upload_images/9902136-78857d903038a84d?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

![IE11-simple.png](http://upload-images.jianshu.io/upload_images/9902136-ff46791595f72bc5?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

![chrome62-more.png](http://upload-images.jianshu.io/upload_images/9902136-4341796d2474eb1f?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

![IE11-more.png](http://upload-images.jianshu.io/upload_images/9902136-de5acf017224cb21?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)

1.2.3 实验链接：

[https://jsperf.com/test-try-catch3](https://jsperf.com/test-try-catch3)

[https://jsperf.com/test-try-catch](https://jsperf.com/test-try-catch)

上面实验数据对比得知，`try catch`会消耗性能，但是`try catch`对`Chrome`的影响比`IE11`小很多,据说是V8引擎新的编译器`TurboFan `起到的作用，有兴趣的小伙伴们可以看下[v8_8h_source的3354行起](https://v8docs.nodesource.com/node-0.8/d4/da0/v8_8h_source.html)，但是`IE11`是slower不少的。这就根据小伙伴们的业务对象了，如果只面向现代浏览器，`try catch`消耗性能影响会很小；如果需要兼容`IE`或内嵌在低端的`webView`时，可适当考虑下`try catch`消耗性能。


##### 2.try catch捕获不到异步错误

尝试对异步方法进行`try catch`操作只能捕获当次事件循环内的异常，对callback执行时抛出的异常将无能为力。

```javascript
try {
    setTimeout(()=>{
        const A = 1
        A = 2
    },0)
} catch (err) {
    // 这里并不能捕获回调里面抛出的异常
    console.log("-----catch error------")
    console.log(err)
}
```
异步情况想捕获异常，建议在异步函数里包一层`try catch`。
```javascript

setTimeout(() => {
  try {
    const A = 1
    A = 2
  } catch (err) {
    console.log(err)
  }
}, 0)

```

##### 3.try catch抛出错误

与 `try-catch` 语句相配的还有一个 `throw` 操作符，随时抛出自定义错误，可以根据不同错误类型，创建自定义错误消息。
```javascript
throw new Error("Something bad happened.");
throw new SyntaxError("I don’t like your syntax.");
throw new TypeError("What type of variable do you take me for?"); throw new RangeError("Sorry, you just don’t have the range.");
throw new EvalError("That doesn’t evaluate.");
throw new URIError("Uri, is that you?");
throw new ReferenceError("You didn’t cite your references properly.");
```
如果觉得自定义的报错不合理，想看原生报错，可以使用`Chrome`的`Pause on exceptions`功能
![image.png](http://upload-images.jianshu.io/upload_images/9902136-e4cbfcb116381944?imageMogr2/auto-orient/strip%7CimageView2/2/w/1240)



### 三、慎用try catch

`try catch`最适合处理那些我们无法控制的错误，如`I/O`操作等，后端`nodeJs`或`java`读取`I/O`操作比较多比如读数据库，所以用`try catch`比较多。前端可以用在上传图片、使用别人的`js`库报错、`async await`同步调接口等地方适用。
``` javascript
async function f() {
  try {
    await Promise.reject('出错了');
  } catch(e) {
  }
  return await Promise.resolve('hello world');
}
```
但是大部分前端客户端代码处理都不怎么依赖环境也没有`I/O`操作，都是自己写的代码，在明明白白地知道自己的代码会发生错误时，再使用`try catch`语句就不太合适了，对应数据类型的错误，建议小伙伴们用解构赋值指定默认值、`&&`和`||`来规避，所以慎用try catch。
```javascript
foo = (obj = {}) => {
  let obj1 = result || {};
  if (obj && obj.code) {
    console.log('obj.code',obj.code)
  }
}
```


### 参考资料

> * [ECMAScript 2015 -The try Statement](https://www.ecma-international.org/ecma-262/6.0/#sec-try-statement-static-semantics-varscopeddeclarations)
> * [https://developers.google.com/web/updates/2015/05/automatically-pause-on-any-exception](https://developers.google.com/web/updates/2015/05/automatically-pause-on-any-exception)
> * [https://v8docs.nodesource.com/node-0.8/d4/da0/v8_8h_source.html](https://v8docs.nodesource.com/node-0.8/d4/da0/v8_8h_source.html)

Happy coding .. :)